File: data_structure/array/73_set_matrix_zeroes.cc
#include <vector>

// 2024/01/05: 自己写出了空间O(mn)的算法. 看答案思路写出了O(1)空间的算法.

// 利用第0行和第0列存储0的信息.
void SetZeroes(std::vector<std::vector<int>>& matrix) {
  bool flag_col0 = false;
  bool flag_row0 = false;
  for (int i = 0; i < matrix.size(); ++i) {
    if (matrix[i][0] == 0) {
      flag_col0 = true;
    }
  }
  for (int j = 0; j < matrix[0].size(); ++j) {
    if (matrix[0][j] == 0) {
      flag_row0 = true;
    }
  }
  for (int i = 1; i < matrix.size(); ++i) {
    for (int j = 1; j < matrix[0].size(); ++j) {
      if (matrix[i][j] == 0) {
        matrix[i][0] = 0;
        matrix[0][j] = 0;
      }
    }
  }
  for (int i = 1; i < matrix.size(); ++i) {
    for (int j = 1; j < matrix[0].size(); ++j) {
      if (matrix[i][0] == 0 || matrix[0][j] == 0) {
        matrix[i][j] = 0;
      }
    }
  }
  if (flag_col0) {
    for (int i = 0; i < matrix.size(); ++i) {
      matrix[i][0] = 0;
    }
  }
  if (flag_row0) {
    for (int j = 0; j < matrix[0].size(); ++j) {
      matrix[0][j] = 0;
    }
  }
}

File: data_structure/array/73_set_matrix_zeroes.py
from typing import List


# 2026/05/11: 自己写出了空间 O(mn) 的算法.
# 空间 O(1), 把矩阵的第 0 行和第 0 列用来存储是否要清零此行(列)的信息.
def set_zeroes(matrix: List[List[int]]) -> None:
    """
    Do not return anything, modify matrix in-place instead.
    """
    if len(matrix) == 0 or len(matrix[0]) == 0:
        return

    row_to_be_clear = set()
    col_to_be_clear = set()
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            if matrix[i][j] == 0:
                row_to_be_clear.add(i)
                col_to_be_clear.add(j)

    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            if i in row_to_be_clear or j in col_to_be_clear:
                matrix[i][j] = 0


matrix1 = [[1, 1, 1], [1, 0, 1], [1, 1, 1]]
matrix2 = [[0, 1, 2, 0], [3, 4, 5, 2], [1, 3, 1, 5]]
set_zeroes(matrix1)
set_zeroes(matrix2)
print(matrix1)
print(matrix2)
